chapter-06/lambdas/create_link/src/main.rs

```diff
@@ -0,0 +1,197 @@
+use lambda_http::{run, service_fn, tracing, Error, IntoResponse, Request, RequestPayloadExt};
+use shared::adapters::DynamoDbUrlRepository;
+use shared::core::{ShortenUrlRequest, UrlShortener};
+use shared::url_info::HttpUrlInfo;
+use shared::utils::generate_api_response;
+use std::env;
+
+async fn function_handler(
+    url_shortener: &UrlShortener,
+    event: Request,
+) -> Result<impl IntoResponse, Error> {
+    tracing::info!("Received event: {:?}", event);
+
+    let shorten_url_request_body = event.payload::<ShortenUrlRequest>()?;
+
+    match shorten_url_request_body {
+        None => generate_api_response(400, "Bad request"),
+        Some(shorten_url_request) => {
+            let shortened_url_response = url_shortener.shorten_url(shorten_url_request).await;
+
+            let response = match shortened_url_response {
+                Ok(response) => {
+                    generate_api_response(200, &serde_json::to_string(&response).unwrap())?
+                }
+                Err(e) => {
+                    tracing::error!("Failed to shorten URL: {:?}", e);
+                    generate_api_response(500, "Internal Server Error")?
+                }
+            };
+
+            Ok(response)
+        }
+    }
+}
+
+#[tokio::main]
+async fn main() -> Result<(), Error> {
+    tracing::init_default_subscriber();
+    let table_name = env::var("TABLE_NAME").expect("TABLE_NAME is not set");
+    let config = aws_config::load_from_env().await;
+    let dynamodb_client = aws_sdk_dynamodb::Client::new(&config);
+    let http_client = shared::Client::builder()
+        .timeout(std::time::Duration::from_secs(2))
+        .build()?;
+    let url_info = HttpUrlInfo::new(http_client);
+    let url_repo = DynamoDbUrlRepository::new(table_name, dynamodb_client);
+    let shortener = UrlShortener::new(Box::new(url_repo), Box::new(url_info));
+
+    run(service_fn(|event| function_handler(&shortener, event))).await
+}
+
+#[cfg(test)]
+mod tests {
+    use async_trait::async_trait;
+    use lambda_http::http::Request;
+    use lambda_http::Body;
+    use lambda_http::IntoResponse;
+    use serde_json::Value;
+    use shared::core::ShortUrl;
+    use shared::core::UrlInfo;
+    use shared::core::UrlRepository;
+    use shared::core::UrlShortener;
+    use shared::url_info::UrlDetails;
+
+    use crate::function_handler;
+
+    #[derive(Debug)]
+    struct TestUrlRepository {
+        return_error: bool,
+    }
+
+    #[async_trait]
+    impl UrlRepository for TestUrlRepository {
```

Comment:
@lmammino for these unit tests I've written for the invidivual functions I can't find a way to use the automocking from `mockall`. Hence the dummy implementations in each folder. Do you know of a better way?

Reply:
I think this is probably simple enough to be ok with a manual implementation. So I'd leave it as it it. It might be good to get the opinion of some experienced Rust devs such as David or Richard...